core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java

```diff
@@ -326,6 +326,85 @@ public void testCastTimestampToTimestampWithTimeZone()
+        }
+
+        List<LocalTime> toLocalTimes = asList(
+                // 2017-09-24 02:59:00 Pacific/Apia is 2017-09-23T13:59:00Z
+                LocalTime.parse("13:59:59.999999999"),
+                // 2017-09-24 04:00:00 Pacific/Apia is 2017-09-23T14:00:00Z
+                // 2017-09-24 03:00:00 gets interpreted as 2017-09-23T14:00:00Z too
+                LocalTime.parse("14:00:00"),
+                LocalTime.parse("14:00:00.000000001"),
+                LocalTime.parse("14:00:00.000000002"),
+                LocalTime.parse("14:59:59.999999999"),
+                LocalTime.parse("15:00:00"),
+                LocalTime.parse("15:00:00.000000001"),
+                LocalTime.parse("15:00:00.000000002"));
+
+        for (LocalTime toLocalTime : toLocalTimes) {
+            for (int timestampPrecision : asList(0, 3, 6, 9, 12)) {
+                for (String operator : COMPARISON_OPERATORS) {
+                    validate(
+                            session,
+                            operator,
+                            "date",
+                            "DATE '2017-09-24'",
+                            format("timestamp(%s) with time zone", max(9, timestampPrecision)),
+                            format("TIMESTAMP '2017-04-01 %s Z'", toLocalTime));
+                }
+            }
+        }
+
+        toLocalTimes = asList(
+                // 2017-04-02 02:59:00 Pacific/Apia is 2017-04-01T12:59:00Z
+                LocalTime.parse("12:59:59.999999999"),
+                // 2017-04-02 03:00:00 Pacific/Apia is 2017-04-01T13:00:00Z
+                LocalTime.parse("13:00:00"),
+                LocalTime.parse("13:00:00.000000001"),
+                LocalTime.parse("13:00:00.000000002"),
+                LocalTime.parse("13:59:59.999999999"),
+                // [2017-04-01T14:00:00Z - 2017-04-01T15:00:00Z) range is not addressable with TIMESTAMP to TIMESTAMP WITH TIME ZONE cast in Pacific/Apia zone
+                LocalTime.parse("14:00:00"),
+                LocalTime.parse("14:00:00.000000001"),
+                LocalTime.parse("14:00:00.000000002"),
+                LocalTime.parse("14:59:59.999999999"),
+                // 2017-04-02 04:00:00 Pacific/Apia is 2017-04-01T15:00:00Z
+                LocalTime.parse("15:00:00"),
+                LocalTime.parse("15:00:00.000000001"),
+                LocalTime.parse("15:00:00.000000002"));
+
+        for (LocalTime toLocalTime : toLocalTimes) {
+            for (int timestampPrecision : asList(0, 3, 6, 9, 12)) {
+                for (String operator : COMPARISON_OPERATORS) {
+                    validate(
+                            session,
+                            operator,
+                            "date",
+                            "DATE '2017-09-24'",
+                            format("timestamp(%s) with time zone", max(9, timestampPrecision)),
+                            format("TIMESTAMP '2017-04-01 %s Z'", toLocalTime));
```

Comment:
where does "2017-04-01" come from?
should we have tests with some other dates on the RHS?

Reply:
Added other dates to RHS. 
The date comes from the exsiting `testCastTimestampToTimestampWithTimeZone`.